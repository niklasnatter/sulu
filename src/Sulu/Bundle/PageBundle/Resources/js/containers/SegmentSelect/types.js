// @flow
export type Value = {[webspaceKey: string]: ?string};
