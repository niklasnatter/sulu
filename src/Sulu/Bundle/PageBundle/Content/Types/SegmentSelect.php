<?php

/*
 * This file is part of Sulu.
 *
 * (c) Sulu GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\PageBundle\Content\Types;

use PHPCR\NodeInterface;
use Sulu\Component\Content\Compat\PropertyInterface;
use Sulu\Component\Content\SimpleContentType;

class SegmentSelect extends SimpleContentType
{
    public function __construct()
    {
        parent::__construct('SegmentSelect', '{}');
    }

    public function write(NodeInterface $node, PropertyInterface $property, $userId, $webspaceKey, $languageCode, $segmentKey)
    {
        parent::write($node, $property, $userId, $webspaceKey, $languageCode, $segmentKey);

        // write a separate property for per webspace to make segment queryable in the smart content data provider
        foreach ($property->getValue() as $webspaceKeyValue => $segmentKeyValue) {
            $webspaceSegmentPropertyName = $property->getName() . '-' . $webspaceKeyValue;
            $node->setProperty($webspaceSegmentPropertyName, $this->removeIllegalCharacters($segmentKeyValue));
        }
    }

    protected function encodeValue($value)
    {
        return \json_encode($value);
    }

    protected function decodeValue($value)
    {
        if (!\is_string($value)) {
            $value = $this->defaultValue;
        }

        return \json_decode($value, true);
    }

    public function exportData($propertyValue)
    {
        return $this->encodeValue($propertyValue);
    }

    public function importData(NodeInterface $node, PropertyInterface $property, $value, $userId, $webspaceKey, $languageCode, $segmentKey = null)
    {
        parent::importData(
            $node,
            $property,
            $this->decodeValue($value),
            $userId,
            $webspaceKey,
            $languageCode,
            $segmentKey
        );
    }
}
