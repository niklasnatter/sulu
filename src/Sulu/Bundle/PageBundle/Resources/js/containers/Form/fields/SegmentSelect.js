// @flow
import React from 'react';
import type {FieldTypeProps} from 'sulu-admin-bundle/types';
import SegmentSelectContainer from '../../SegmentSelect';
import type {Value} from '../../SegmentSelect/types';

export default class SegmentSelect extends React.Component<FieldTypeProps<Value>> {
    handleChange = (value: Value) => {
        const {onChange, onFinish} = this.props;

        onChange(value);
        onFinish();
    };

    render() {
        const {disabled, formInspector, value} = this.props;

        return (
            <SegmentSelectContainer
                disabled={disabled}
                onChange={this.handleChange}
                value={value}
                webspace={formInspector.metadataOptions?.webspace}
            />
        );
    }
}
