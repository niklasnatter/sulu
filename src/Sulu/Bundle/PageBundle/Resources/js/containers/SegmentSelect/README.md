This component shows selects for the available segments. If the component is used for a specific webspace, then only a
single Select is shown, otherwise it shows a separate select for every webspace.
