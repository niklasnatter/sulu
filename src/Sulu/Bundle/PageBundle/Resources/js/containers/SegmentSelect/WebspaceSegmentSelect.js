// @flow
import React from 'react';
import {SingleSelect} from 'sulu-admin-bundle/components';
import fieldStyles from 'sulu-admin-bundle/components/Form/field.scss';
import {translate} from 'sulu-admin-bundle/utils';
import type {Webspace} from '../../stores/webspaceStore/types';
import webspaceSegmentSelectStyles from './webspaceSegmentSelect.scss';

export default class WebspaceSegmentSelect extends React.Component<{
    disabled: boolean,
    onChange: (webspaceKey: string, segment: ?string) => void,
    value: ?string,
    webspace: Webspace,
}> {
    handleSelectChange = (value: ?string) => {
        const {onChange, webspace} = this.props;

        onChange(webspace.key, value);
    };

    render() {
        const {disabled, value, webspace} = this.props;

        return (
            <div className={webspaceSegmentSelectStyles.webspaceSection}>
                <label className={fieldStyles.label}>
                    {webspace.name} - {translate('sulu_admin.segments')}
                </label>

                <SingleSelect
                    disabled={!!disabled}
                    onChange={this.handleSelectChange}
                    value={value}
                >
                    <SingleSelect.Option>
                        {translate('sulu_admin.none_selected')}
                    </SingleSelect.Option>
                    {webspace.segments.map(({key, title}) => (
                        <SingleSelect.Option key={key} value={key}>
                            {title}
                        </SingleSelect.Option>
                    ))}
                </SingleSelect>
            </div>
        );
    }
}
