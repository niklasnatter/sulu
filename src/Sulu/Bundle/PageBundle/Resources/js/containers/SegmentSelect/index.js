// @flow
import SegmentSelect from './SegmentSelect';

export default SegmentSelect;
