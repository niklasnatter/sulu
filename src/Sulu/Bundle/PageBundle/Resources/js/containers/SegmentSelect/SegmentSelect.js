// @flow
import React, {Fragment} from 'react';
import {computed} from 'mobx';
import webspaceStore from '../../stores/webspaceStore';
import type {Webspace} from '../../stores/webspaceStore/types';
import WebspaceSegmentSelect from './WebspaceSegmentSelect';
import type {Value} from './types';

type Props = {|
    disabled: ?boolean,
    onChange: (value: Value) => void,
    value: ?Value,
    webspace: ?string,
|};

export default class SegmentSelect extends React.Component<Props> {
    @computed get visibleWebspaces(): Array<Webspace> {
        const {webspace} = this.props;

        // we want to select a segment for each webspace on webspace-independent objects like articles
        const matchingWebspaces: Array<Webspace> = webspace
            ? [webspaceStore.getWebspace(webspace)]
            : webspaceStore.grantedWebspaces;

        return matchingWebspaces.filter((webspace) => webspace.segments.length > 0);
    }

    handleWebspaceSegmentChange = (webspaceKey: string, segment: ?string) => {
        const {onChange, value} = this.props;

        onChange({...value, [webspaceKey]: segment});
    };

    render() {
        const {disabled, value} = this.props;

        return (
            <Fragment>
                {this.visibleWebspaces.map((webspace) => {
                    return (
                        <WebspaceSegmentSelect
                            disabled={!!disabled}
                            key={webspace.key}
                            onChange={this.handleWebspaceSegmentChange}
                            value={value ? value[webspace.key] : undefined}
                            webspace={webspace}
                            webspaceNameVisible={this.visibleWebspaces.length > 1}
                        />
                    );
                })}
            </Fragment>
        );
    }
}
