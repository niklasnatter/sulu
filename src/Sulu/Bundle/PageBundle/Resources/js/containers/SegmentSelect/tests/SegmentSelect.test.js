// @flow
import React from 'react';
import {mount} from 'enzyme';
import webspaceStore from '../../../stores/webspaceStore';
import SegmentSelect from '../../SegmentSelect';

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('../../../stores/webspaceStore', () => ({
    getWebspace: jest.fn(),
}));

test('Render a label and a SingleSelect for each granted webspace that has segments', () => {
    // $FlowFixMe
    webspaceStore.grantedWebspaces = [
        {
            name: 'Webspace One',
            key: 'webspace-1',
            segments: [
                {key: 'w', title: 'Winter'},
                {key: 's', title: 'Summer'},
            ],
        },
        {
            name: 'Webspace Two',
            key: 'webspace-2',
            segments: [],
        },
        {
            name: 'Webspace Three',
            key: 'webspace-3',
            segments: [
                {key: 'a', title: 'Autumn'},
                {key: 'p', title: 'Spring'},
            ],
        },
    ];

    const segmentSelect = mount(
        <SegmentSelect
            disabled={false}
            onChange={jest.fn()}
            value={{
                'webspace-1': 's',
            }}
            webspace={undefined}
        />
    );

    expect(segmentSelect.find('SingleSelect')).toHaveLength(2);
    expect(segmentSelect.render()).toMatchSnapshot();
});

test('Render a label without webspace name if only one webspace has segments', () => {
    // $FlowFixMe
    webspaceStore.grantedWebspaces = [
        {
            name: 'Webspace One',
            key: 'webspace-1',
            segments: [
                {key: 'w', title: 'Winter'},
                {key: 's', title: 'Summer'},
            ],
        },
    ];

    const segmentSelect = mount(
        <SegmentSelect
            disabled={false}
            onChange={jest.fn()}
            value={{
                'webspace-1': 's',
            }}
            webspace={undefined}
        />
    );

    expect(segmentSelect.find('label')).toHaveLength(1);
    expect(segmentSelect.find('label').text()).toEqual('sulu_admin.segment');
});

test('Render a label and a SingleSelect for each granted webspace that has segments', () => {
    // $FlowFixMe
    webspaceStore.grantedWebspaces = [
        {
            name: 'Webspace One',
            key: 'webspace-1',
            segments: [
                {key: 'w', title: 'Winter'},
                {key: 's', title: 'Summer'},
            ],
        },
        {
            name: 'Webspace Two',
            key: 'webspace-2',
            segments: [],
        },
        {
            name: 'Webspace Three',
            key: 'webspace-3',
            segments: [
                {key: 'a', title: 'Autumn'},
                {key: 'p', title: 'Spring'},
            ],
        },
    ];

    const segmentSelect = mount(
        <SegmentSelect
            disabled={false}
            onChange={jest.fn()}
            value={{
                'webspace-1': 's',
            }}
            webspace={undefined}
        />
    );

    expect(segmentSelect.find('SingleSelect')).toHaveLength(2);
    expect(segmentSelect.render()).toMatchSnapshot();
});

test('Render only one label and SingleSelect if options contain a webspace', () => {
    const webspace = {
        name: 'Webspace One',
        key: 'webspace-1',
        segments: [
            {key: 'w', title: 'Winter'},
            {key: 's', title: 'Summer'},
        ],
    };
    webspaceStore.getWebspace.mockReturnValue(webspace);

    const segmentSelect = mount(
        <SegmentSelect
            disabled={false}
            onChange={jest.fn()}
            value={{}}
            webspace="webspace-1"
        />
    );

    expect(webspaceStore.getWebspace).toBeCalledWith('webspace-1');
    expect(segmentSelect.find('SingleSelect')).toHaveLength(1);
});

test('Pass correct props to SingleSelect', () => {
    // $FlowFixMe
    webspaceStore.grantedWebspaces = [
        {
            name: 'Webspace One',
            key: 'webspace-1',
            segments: [
                {key: 'w', title: 'Winter'},
                {key: 's', title: 'Summer'},
            ],
        },
        {
            name: 'Webspace Two',
            key: 'webspace-2',
            segments: [],
        },
        {
            name: 'Webspace Three',
            key: 'webspace-3',
            segments: [
                {key: 'a', title: 'Autumn'},
                {key: 'p', title: 'Spring'},
            ],
        },
    ];

    const segmentSelect = mount(
        <SegmentSelect
            disabled={true}
            onChange={jest.fn()}
            value={{
                'webspace-1': 's',
            }}
            webspace={undefined}
        />
    );

    segmentSelect.find('Select').at(0).instance().openOptionList();
    segmentSelect.update();

    expect(segmentSelect.find('SingleSelect').at(0).prop('disabled')).toEqual(true);
    expect(segmentSelect.find('SingleSelect').at(0).prop('value')).toEqual('s');
    expect(segmentSelect.find('SingleSelect').at(1).prop('disabled')).toEqual(true);
    expect(segmentSelect.find('SingleSelect').at(1).prop('value')).toEqual(undefined);

    expect(segmentSelect.find('Option').at(0).prop('children')).toEqual('sulu_admin.none_selected');
    expect(segmentSelect.find('Option').at(0).prop('value')).toEqual(undefined);
    expect(segmentSelect.find('Option').at(1).prop('children')).toEqual('Winter');
    expect(segmentSelect.find('Option').at(1).prop('value')).toEqual('w');
    expect(segmentSelect.find('Option').at(2).prop('children')).toEqual('Summer');
    expect(segmentSelect.find('Option').at(2).prop('value')).toEqual('s');
});

test('Call onChange if the value is changed', () => {
    const changeSpy = jest.fn();

    // $FlowFixMe
    webspaceStore.grantedWebspaces = [
        {
            name: 'Webspace One',
            key: 'webspace-1',
            segments: [
                {key: 'w', title: 'Winter'},
                {key: 's', title: 'Summer'},
            ],
        },
        {
            name: 'Webspace Two',
            key: 'webspace-2',
            segments: [],
        },
        {
            name: 'Webspace Three',
            key: 'webspace-3',
            segments: [
                {key: 'a', title: 'Autumn'},
                {key: 'p', title: 'Spring'},
            ],
        },
    ];

    const segmentSelect = mount(
        <SegmentSelect
            disabled={true}
            onChange={changeSpy}
            value={{
                'webspace-1': 's',
            }}
            webspace={undefined}
        />
    );

    segmentSelect.find('SingleSelect').at(1).prop('onChange')('a');
    expect(changeSpy).toBeCalledWith({
        'webspace-1': 's',
        'webspace-3': 'a',
    });
});
