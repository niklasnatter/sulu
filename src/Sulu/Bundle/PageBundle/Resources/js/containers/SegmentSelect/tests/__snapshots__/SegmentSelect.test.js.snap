// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a label and a SingleSelect for each granted webspace that has segments 1`] = `
Array [
  <div
    class="webspaceSection"
  >
    <label
      class="label"
    >
      Webspace One - sulu_admin.segments
    </label>
    <div
      class="select"
    >
      <button
        class="displayValue default"
        type="button"
      >
        <div
          aria-label="Summer"
          class="croppedText"
          title="Summer"
        >
          <div
            aria-hidden="true"
            class="front"
          >
            Sum
          </div>
          <div
            aria-hidden="true"
            class="back"
          >
            <span>
              mer
            </span>
          </div>
          <div
            class="whole"
          >
            Summer
          </div>
        </div>
        <span
          aria-label="su-angle-down"
          class="su-angle-down toggle"
        />
      </button>
    </div>
  </div>,
  <div
    class="webspaceSection"
  >
    <label
      class="label"
    >
      Webspace Three - sulu_admin.segments
    </label>
    <div
      class="select"
    >
      <button
        class="displayValue default"
        type="button"
      >
        <div
          aria-label="sulu_admin.none_selected"
          class="croppedText"
          title="sulu_admin.none_selected"
        >
          <div
            aria-hidden="true"
            class="front"
          >
            sulu_admin.n
          </div>
          <div
            aria-hidden="true"
            class="back"
          >
            <span>
              one_selected
            </span>
          </div>
          <div
            class="whole"
          >
            sulu_admin.none_selected
          </div>
        </div>
        <span
          aria-label="su-angle-down"
          class="su-angle-down toggle"
        />
      </button>
    </div>
  </div>,
]
`;
